From khali@linux-fr.org Fri Sep 23 00:05:12 2005
Date: Thu, 22 Sep 2005 22:15:53 +0200
From: Jean Delvare <khali@linux-fr.org>
To: Greg KH <greg@kroah.com>
Subject: [PATCH] (6/7) i2c-viapro: Refactor control outb
Message-Id: <20050922221553.1a51a3a0.khali@linux-fr.org>

By slightly shifting the interface between vt596_access and
vt596_transaction, we can save two I/O accesses per SMBus transaction.

Signed-off-by: Jean Delvare <khali@linux-fr.org>
Signed-off-by: Greg Kroah-Hartman <gregkh@suse.de>


 drivers/i2c/busses/i2c-viapro.c |    7 +++----
 1 file changed, 3 insertions(+), 4 deletions(-)

---
--- gregkh-2.6.orig/drivers/i2c/busses/i2c-viapro.c
+++ gregkh-2.6/drivers/i2c/busses/i2c-viapro.c
@@ -105,7 +105,7 @@ static struct i2c_adapter vt596_adapter;
 static unsigned int vt596_features;
 
 /* Return -1 on error, 0 on success */
-static int vt596_transaction(void)
+static int vt596_transaction(u8 size)
 {
 	int temp;
 	int result = 0;
@@ -131,7 +131,7 @@ static int vt596_transaction(void)
 	}
 
 	/* Start the transaction by setting bit 6 */
-	outb_p(inb(SMBHSTCNT) | 0x40, SMBHSTCNT);
+	outb_p(0x40 | (size & 0x3C), SMBHSTCNT);
 
 	/* We will always wait for a fraction of a second */
 	do {
@@ -232,9 +232,8 @@ static s32 vt596_access(struct i2c_adapt
 	}
 
 	outb_p(((addr & 0x7f) << 1) | read_write, SMBHSTADD);
-	outb_p((size & 0x3C), SMBHSTCNT);
 
-	if (vt596_transaction()) /* Error in transaction */
+	if (vt596_transaction(size)) /* Error in transaction */
 		return -1;
 
 	if ((read_write == I2C_SMBUS_WRITE) || (size == VT596_QUICK))
